pyflakes/test/test_type_annotations.py

```diff
@@ -667,3 +667,20 @@ class C(Protocol):
                 def f():  # type: () -> int
                     pass
         """)
+
+    def test_recursive_global(self):
+        self.flakes("""
+            from typing import Optional, Tuple
+
+            class NestedTuple:
+                x: Tuple[int, Optional['NestedTuple']]
+        """)
+
+    def test_recursive_local(self):
```

Comment:
This used to fail because on the first check the ast just has the "function def" node
So when checking the inside, `self._in_deferred` was now set to `True` causing the issue.
I don't really know why we can't just run everything without it. I remove the  `_in_deferred` boolean and all the unit tests work. I also checked locally on different projects without any issue.
If this causes other problems, I'll be happy to add an extra unit test for this and work on another way to fix it